packages/tax/src/services/tax-module-service.ts

```diff
@@ -166,4 +167,145 @@ export default class TaxModuleService<
   ) {
     return await this.taxRateRuleService_.create(data, sharedContext)
   }
+
+  async getTaxLines(
+    items: (TaxTypes.TaxableItemDTO | TaxTypes.TaxableShippingDTO)[],
+    calculationContext: TaxTypes.TaxCalculationContext,
+    @MedusaContext() sharedContext: Context = {}
+  ): Promise<(TaxTypes.ItemTaxLineDTO | TaxTypes.ShippingTaxLineDTO)[]> {
+    const regions = await this.taxRegionService_.list(
+      {
+        $or: [
+          {
+            country_code: calculationContext.address.country_code,
+            province_code: null,
+          },
+          {
+            country_code: calculationContext.address.country_code,
+            province_code: calculationContext.address.province_code,
+          },
+        ],
+      },
+      {},
+      sharedContext
+    )
+
+    const toReturn = await promiseAll(
+      items.map(async (item) => {
+        const isShipping = "shipping_option_id" in item
+        let ruleQuery = isShipping
+          ? [
+              {
+                reference: "shipping_option",
+                reference_id: item.shipping_option_id,
+              },
+            ]
+          : [
+              {
+                reference: "product",
+                reference_id: item.product_id,
+              },
+              {
+                reference: "product_type",
+                reference_id: item.product_type_id,
+              },
+            ]
+
+        const rates = await this.taxRateService_.list(
+          {
+            $and: [
+              { tax_region_id: { $in: regions.map((r) => r.id) } },
+              { $or: [{ is_default: true }, { rules: { $or: ruleQuery } }] },
+            ],
+          },
+          { relations: ["rules", "tax_region"] },
+          sharedContext
+        )
+
+        const prioritizedRates = this.prioritizeRates(rates, item)
+
+        const rate = prioritizedRates[0]
+        const toReturn = {
+          rate_id: rate.id,
+          rate: rate.rate,
+          code: rate.code,
+          name: rate.name,
+        }
+
+        if (isShipping) {
+          return [{ shipping_line_id: item.id, ...toReturn }]
+        }
+
+        return [{ line_item_id: item.id, ...toReturn }]
+      })
+    )
+
+    return toReturn.flat()
+  }
+
+  private prioritizeRates(
```

Comment:
Is it possible to have something a bit more generic here?
Something a bit like this:

```typescript
const taxProperties = {
  province_code: { weight: 50 },
  is_default: { weight: 10 },
}

const itemProperties = {
  product_id: { weight: 100 },
  product_type_id: { weight: 50 },
  shipping_option_id: { weight: 100 },
  // any other item property
}

function prioritizeRates(rates, item) {
  const calculateWeight = (rate) => {
    let weight = rate.is_default ? itemProperties.is_default.weight : 0

    for (const property in taxProperties) {
      if (rate.tax_region[property] === null) {
        continue
      }

      const ruleConfig = taxProperties[property]
      weight += ruleConfig.weight
    }

    for (const rule of rate.rules) {
      const itemProperty = `${rule.reference}_id`
      if (!(itemProperty in itemProperties) || !(itemProperty in item)) {
        continue
      }

      if (rule.reference_id === item[itemProperty]) {
        const ruleConfig = itemProperties[itemProperty]
        weight += ruleConfig.weight
      }
    }

    return weight
  }

  return rates
    .map((rate) => ({ ...rate, weight: calculateWeight(rate) }))
    .sort((a, b) => b.weight - a.weight) // weight DESC
}
```
forgive any mistake here, it was a quick implementation of an idea. If not applicable, just disregard it.

Reply:
Love this idea!

Reply:
After playing around with it I think I will stick with what's there for now. Reason being:
1. It's tricky to figure out what the exact priority of combinations are with this.
2. Right now we pull out all rules for an override rate (which might be a lot), this should change to only pull out the rates relevant to the item. This will simplify the current model more and maybe make this approach less relevant.